### PR TITLE
🌱 add check to ensure ClusterClass change doesn't invalidate existing cluster variables

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

 This PR adds a check to ensure variables which are currently in use by clusters are not invalidated when a clusterClass is updated.

Fixes: #5461

### CODE REVIEW

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 1]
```diff
@@ -95,6 +98,21 @@ func validateClusterClassVariableName(variableName string, fldPath *field.Path)
 	return allErrs
 }
 
+// validateRequiredVariablesAreDefaulted validates that ClusterClass variable names are unique.
+func validateRequiredVariablesAreDefaulted(clusterClassVariable *clusterv1.ClusterClassVariable, pathPrefix *field.Path) field.ErrorList {
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:102]: I don't think we should enforce that **all** required variables **must** have default values.

This only applies to variables newly added during a ClusterClass update, because the corresponding Clusters cannot have values for those variables.

In case of a new ClusterClass default values are not required. I think there are cases of required variables where there is simply no valid default value (e.g. azure subscription id: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template.yaml#L44)

Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation.go:102]: I think we should make this part of our regular "does the new schema match on all Clusters" validation code. This way new required variables can be introduced like this by the user:

* add new variable with required false (aka optional) to the ClusterClass (existence of default value does not matter)
* add variable values to all Clusters
* make the variable required in the ClusterClass

With this kind of validation we can even account for changes between required true/false without any additional code for that special case.

File: internal/webhooks/clusterclass.go [hunk 2]
```diff
@@ -154,17 +155,66 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+		variables.ValidateClusterClassVariables(new.Spec.Variables, field.NewPath("spec", "variables"))...,
+	)
 
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
 
+func (webhook *ClusterClass) validateVariablesNotInvalidated(clusters []clusterv1.Cluster, old, new *clusterv1.ClusterClass) field.ErrorList {
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:176]: I think this func has to work slightly differently:

* loop over all Variables in the ClusterClass and find out:
    * which variables have been removed
    * which variables have been changed (including required true/false & schema changes)
    * which variables have been added

Then we have to do the following:

* for removed variables:
    *  make sure they don't exist in any Cluster
* for changed variables:
    * if required: make sure the variable exist on all clusters and adheres to the schema
    * if not required: if the variable exist on a cluster it must adhere to the scheme, but it's okay if it doesn't exist
* for added variables:
    * if required: make sure the variable exist on all clusters and adheres to the schema
    * if not required: if the variable exist on a cluster it must adhere to the scheme, but it's okay if it doesn't exist
    
 By also handling "added variables" here we don't need `validateRequiredVariablesAreDefaulted` in `validateClusterClassVariable` anymore.

**Review by sbueringer — COMMENTED** [review]
high level review for now

File: internal/topology/variables/clusterclass_variable_validation.go (hunk 1 above)
Comment (fabriziopandini) [reply internal/topology/variables/clusterclass_variable_validation.go:102]: +1 to make this part of our regular "does the new schema match on all Clusters" validation code.

File: internal/webhooks/clusterclass.go [hunk 3]
```diff
@@ -168,6 +176,147 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+
+	// Get the old ClusterClassVariables as a map
+	oldVars, _ := getClusterClassVariablesMapWithReverseIndex(old.Spec.Variables)
+
+	// Get the old ClusterClassVariables as a map with an index linking them to their place in the ClusterClass Variable array.
+	// Note: The index is used to improve the error recording below.
+	newVars, clusterClassVariablesIndex := getClusterClassVariablesMapWithReverseIndex(new.Spec.Variables)
+
+	// Retrieve the set of ClusterClass variables that require validation.
+	varsForValidation := getClusterClassVariablesForValidation(oldVars, newVars)
```
Comment (fabriziopandini) [inline internal/webhooks/clusterclass.go:198]: What about
```suggestion
	// Compute the diff between old an new variables.
	varsDiff := getClusterClassVariablesDiff(oldVars, newVars)
```

File: internal/webhooks/clusterclass.go [hunk 4]
```diff
@@ -168,6 +176,147 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+
+	// Validate the ClusterClassVariables for every ClusterClassVariable on every Cluster using the ClusterClass.
```
Comment (fabriziopandini) [inline internal/webhooks/clusterclass.go:200]:
```suggestion
	// Validate the variable values on each Cluster ensuring they are still compatible with the new
	// schemas defined in the ClusterClass
```

File: internal/webhooks/clusterclass.go [hunk 5]
```diff
@@ -168,6 +176,147 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+			// Add Cluster Variable entry in clusterVariableReferences to track where it is in use.
+			if _, ok := clusterVariableReferences[clusterVar.Name]; !ok {
+				clusterVariableReferences[clusterVar.Name] = []string{}
+			}
+			clusterVariableReferences[clusterVar.Name] = append(clusterVariableReferences[clusterVar.Name], cluster.Name)
+
+			// 1) Error if a variable with a schema altered in the update is no longer valid on the Cluster.
+			if alteredVar, ok := varsForValidation.altered[clusterVar.Name]; ok {
+				allErrs = append(allErrs, variables.ValidateClusterVariable(clusterVar.DeepCopy(), alteredVar, path.Index(varIndex),
+					fmt.Sprintf("variable %s in Cluster %s is no longer valid", alteredVar.Name, cluster.Name))...)
```
Comment (fabriziopandini) [inline internal/webhooks/clusterclass.go:218]: Assuming that a ClusterClass is used by many Cluster, this approach can lead to many errors.

What about maintaining a map[variableName][]ClusterName where we store for each variable the Clusters that get invalidated by the change.

Then we can return an error like "changes to the %s variable definition cannot be applied because they are incompatible with values defined in the following Cluster: %s" or something similar.
Note: I understand we are losing the error from ValidateClusterVariable, but this is a trade-off, but eventually in future iterations we can make them surface avoiding duplications
Same for other errors

File: internal/webhooks/clusterclass.go [hunk 6]
```diff
@@ -168,6 +176,147 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+				allErrs = append(allErrs,
+					variables.ValidateClusterVariable(clusterVar.DeepCopy(), addedVar, path.Index(varIndex),
+						fmt.Sprintf("variable %s in Cluster %s is no longer valid", addedVar.Name, cluster.Name))...)
+			}
+		}
+	}
+
+	// 4) Error if a required variable is not defined in every cluster using the ClusterClass.
+	for varName := range varsForValidation.required {
+		if len(clusterVariableReferences[varName]) != len(clusters) {
```
Comment (fabriziopandini) [inline internal/webhooks/clusterclass.go:243]: In order to make this consistent with the above, I will loop on all clusters and keep track of the Cluster where the variable is missing (while instead now we are keeping track of the Clusters where the variable is defined)

File: internal/webhooks/clusterclass.go (hunk 5 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: I've implemented a solution for this in the new version  - let me know what you think.

File: internal/webhooks/clusterclass.go [hunk 7]
```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable %s can not be removed. It is used in Clusters %v", variable, clusters),
+				))
+		}
+	}
+	if len(errs.altered) > 0 {
+		for variable, clusters := range errs.altered {
+			allErrs = append(allErrs,
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable schema for %s can not be altered. It failed validation in Clusters %v", variable, clusters),
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:271]:
```suggestion
					fmt.Sprintf("variable schema change for variable %s is invalid. It failed validation in Clusters %v", variable, clusters),
```
nit. Just to avoid confused users which might now think that the schema cannot be altered (as it can be altered just not to *that* schema)

File: internal/webhooks/clusterclass.go [hunk 8]
```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+			errorInfo.required.add(varName, clustersMissingVariable...)
+		}
+	}
+	// Aggregate the errors from the validation checks and return one error message of each type for each variable with a list of violating clusters.
+
+	return aggregateErrors(errorInfo, clusterClassVariablesIndex, path)
+}
+
+// errorAggregator used to aggregate information about errors during variable comparison.
+// TODO(killianmuldoon): Make the errorAggregator generic so it can be used in cases where a single error aggregates errors related to many objects.
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:248]: I have a hard time understanding how the current implementation can be evolved to something generic as it seems to be very coupled to the ideas of:
* altered / added / removed / required variables
* field errors

Seems like we have to abstract away:
* specific "error states" (altered, added,...) or implement the generic aggregator only for one of them and then use it multiple times here
* the error creation (including error message generation and field error generation)

Sounds doable, but I'm not sure how useful / reusable the result would be. But something for a potential other PR/issue, right?

File: internal/webhooks/clusterclass_test.go [hunk 9]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
 	}
 }
 
+func TestClusterClassValidationWithVariableChecks(t *testing.T) {
+	// NOTE: ClusterTopology feature flag is disabled by default, thus preventing to create or update ClusterClasses.
+	// Enabling the feature flag temporarily for this test.
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+	clusterClassBuilder := builder.ClusterClass(metav1.NamespaceDefault, "class1").
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1328]: reusing the clusterClassBuilder only works because we always overwrite the same field, right? (otherwise we would collect fields from one test case to another).

WDYT about the alternative of just using arrays of ClusterClassVariables in the table and then building the full ClusterClass in the for loop? (should also safe as ~4 lines per test case, but that's not my primary reason :))

File: internal/webhooks/clusterclass_test.go [hunk 10]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+			builder.ControlPlaneTemplate(metav1.NamespaceDefault, "cp1").
+				Build())
+	tests := []struct {
+		name            string
+		oldClusterClass *clusterv1.ClusterClass
+		newClusterClass *clusterv1.ClusterClass
+		clusters        []client.Object
+		expectErr       bool
+	}{
+		{
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1341]: NOTE: nullables are gone on main, so I think you can simplify a bit by rebasing

File: internal/webhooks/clusterclass_test.go [hunk 11]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+									{Raw: []byte(`"us-east-3"`)},
+								},
+							},
+						},
+					}).
+				Build(),
+			expectErr: true,
+		},
+		{
+			name: "Pass if a Minimum ClusterClassVariable is changed but does not invalidate an existing ClusterVariable",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1535]: I think that's the same as the first case

Comment (sbueringer) [reply internal/webhooks/clusterclass_test.go:1535]: Similar for the next test

File: internal/webhooks/clusterclass_test.go [hunk 12]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+								// Type changed to a value that does not invalidate ClusterVariable.
+								Type: "number",
+							},
+						},
+					}).
+				Build(),
+			expectErr: false,
+		},
+		{
+			name: "Pass if Type ClusterClassVariable is changed ( Number-> Integer) not invalidating an existing ClusterVariable",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1807]: Looks like this does not pass (as expected), so I think the name should be adjusted accordingly.

File: internal/webhooks/clusterclass_test.go [hunk 13]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+							OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+								Type: "integer",
+							},
+						},
+					}).
+				Build(),
+			expectErr: true,
+		},
+		{
+			name: "Error if update to make variable required is rejected if not defined in all clusters",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:2034]: This reads slightly confusing. Is it essentially the same as the case above with more clusters/variables?

File: internal/webhooks/clusterclass_test.go [hunk 14]
```diff
@@ -1318,6 +1320,854 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+						Required: true,
+						Schema: clusterv1.VariableSchema{
+							OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+								Type: "number",
+							},
+						},
+					},
+				).
+				Build(),
+			expectErr: true,
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:2146]: Maybe some comments in between to make it easier to navigate would be good (we can refer to which of the validations 1-4 we're testing in the respective test blocks.

Should we add tests for: (I could have missed some existing tests)
* `3)`
    * test that it's fine to add a new optional variable
    * test that we get an error if an optional variable is added and it exists on a cluster but the schema does not match

I think everything else is covered, but maybe you can just double-check to make sure we're not missing a case.

**Review by sbueringer — COMMENTED** [review]
Looks great!

I really like how this validation implementation allows users to do very flexible schema changes safely!!

I think it will also scale very nicely to objects and arrays. I expect that we only have to extend the `schemasAreEqual` func and add more unit tests to cover those new cases.


One downside of the current error aggregation is that users won't see the schema validation errors, so they only know that they don't match and in which clusters but they have to figure out by themselves what exactly is wrong.
But I think it's a fair trade-off to make to keep the validation errors readable/concise, especially on the first iteration.

File: internal/topology/variables/clusterclass_variable_validation.go (hunk 1 above)
Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation.go]: Can be resolved.

File: internal/webhooks/clusterclass.go (hunk 2 above)
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Can be resolved.

File: internal/webhooks/clusterclass.go (hunk 5 above)
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Looks good to me!

File: internal/webhooks/clusterclass_test.go (hunk 9 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass_test.go:1328]: I'm going to do a full review / rewrite of all the validation test cases, there's a lot of repetition and redundancy due to the piece by piece nature of the implementation. I'll push this to a future PR.

File: internal/webhooks/clusterclass.go (hunk 8 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:248]: Yeah - I think there's a lot to think about on this one - I'll create an issue for it.

File: internal/topology/variables/clusterclass_variable_validation_test.go [hunk 15]
```diff
@@ -36,8 +36,7 @@ func Test_ValidateClusterClassVariables(t *testing.T) {
 			name: "Error if multiple variables share a name",
 			clusterClassVariables: []clusterv1.ClusterClassVariable{
 				{
-					Name:     "cpu",
-					Required: true,
+					Name: "cpu",
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation_test.go:39]: Okay that's fine :). I thought something like that, but wasn't 100% sure.

File: internal/webhooks/clusterclass.go [hunk 16]
```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
 	return nil
 }
 
+// validateVariableUpdates checks to see what updates have been made to ClusterClassVariables in the update.
+// It validates:
+// 1) Altered Variable definitions on any exiting Cluster are still valid with the updated Schema.
+// 2) Removed Variables are not in use on any Cluster using the ClusterClass.
+// 3) Required Variables are defined on each Cluster using the ClusterClass.
+// 4) Add Variable definitions on any exiting Cluster are still valid with the updated Schema.
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:184]: I think this loses the structure of the original comment - but there's definite clarification to be done. I'll try to merge the two approaches 😄

File: internal/webhooks/clusterclass.go [hunk 17]
```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+		varClusters[cluster] = nil
+	}
+	var clusterDiff []string
+	for _, cluster := range allClusters {
+		if _, found := varClusters[cluster]; !found {
+			clusterDiff = append(clusterDiff, cluster)
+		}
+	}
+	return clusterDiff
+}
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:306]: I think using map[string]interface{} as a set is an established enough pattern in Go - but if preferred I could use the k8s set here or a map[string]bool.

File: internal/webhooks/clusterclass.go [hunk 18]
```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+
+	return clusterClassVariablesForValidation{
+		altered:  alteredVariables,
+		added:    addedVariables,
+		removed:  removedVariables,
+		required: requiredVariables,
+	}
+}
+
+func schemasAreEqual(old, new clusterv1.VariableSchema) bool {
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:373]: That's the problem with this func. I wonder if there's some test I can put together to keep this in sync (or we could add this method in the API package - possibly a generated deepequal?). WDYT?

[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:373]: I think we can punt for now and create an issue though.

File: internal/webhooks/clusterclass_test.go (hunk 13 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass_test.go:2034]: Yeah - I'll add a comment, but I think it's good to have a test case that does a little bit of stretching of our array etc. logic.

File: internal/webhooks/clusterclass.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Not sure. I usually see alternatively `map[string]struct{}` and then the value is set to `struct{}{}`

I wasn't actually sure it works by setting the value to `nil` until I checked it with the debugger.

But as it works, I can absolutely life with keep using the same map, the other var names might be a good improvement independently though.

File: internal/webhooks/clusterclass.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go:371]: I think one way to fix it in the prod code is to use `reflect.DeepEqual` (IIRC that should also work).

Not sure about a generated func. I think we would have to write a `DeepEqual` generator for that, which is probably too much effort.

For the current problem I added the same note on my PR. I think for this time it's enough to keep those two notes open and then fix it on whichever PR merges later (but I'm also fine with an issue of course)

File: internal/webhooks/clusterclass_test.go (hunk 13 above)
Comment (sbueringer) [reply internal/webhooks/clusterclass_test.go:2006]: Yup that's fine. It's just that this name is constructed relatively complicated compared to the one above. (error and rejected sounds redundant and makes it (relatively) hard to parse)

File: internal/webhooks/clusterclass.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:371]: reflect.DeepEqual doesn't work for the schema because it includes pointers, but I guess it will work in objects as long as there's no pointers?

[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go:371]: I got it to work in my object/array PR: https://github.com/kubernetes-sigs/cluster-api/pull/5779/files with all kinds of schemas.

Maybe I'm missing something, but maybe the cause was something else the last time when it didn't work.

File: internal/webhooks/clusterclass.go [hunk 19]
```diff
@@ -154,10 +155,17 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
 		// Ensure no MachineDeploymentClass currently in use has been removed from the ClusterClass.
-		allErrs = append(allErrs, webhook.validateRemovedMachineDeploymentClassesAreNotUsed(clusters, old, new)...)
+		allErrs = append(allErrs,
+			webhook.validateRemovedMachineDeploymentClassesAreNotUsed(clusters, old, new)...)
+
+		// Ensure no Variable would be invalidated by the update in spec
+		allErrs = append(allErrs,
+			validateVariableUpdates(clusters, old, new, field.NewPath("spec", "variables"))...)
 	}
 	// Validate variables.
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: :+1: to the move. That leaves only the question open if we should return early in that case. Both options are fine for me. Let's see what others think

File: internal/webhooks/clusterclass.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go:371]: I can't see what the difference would be here - but replacing this method with reflect.DeepEqual fails for the pointer values. Need to do some digging about what the difference is.

File: internal/webhooks/clusterclass.go [hunk 20]
```diff
@@ -154,20 +163,226 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
 
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
 
+// validateVariableUpdates checks if the updates made to ClusterClassVariables are valid.
+// It retrieves a list of variables of interest and validates:
+// 1) Altered Variables defined on any exiting Cluster are still valid with the updated Schema.
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:182]: The gap between godoc and implementation is that the variables weren't altered (or added in 4.), their corresponding ClusterClassVariable / schema was altered / added.

But I don't want to nitpick to much here, but I think as this func is fairly complicated it would be good that readers get a clear picture. Up to you :)

File: internal/webhooks/clusterclass_test.go [hunk 21]
```diff
@@ -1318,6 +1320,824 @@ func TestClusterClassValidationWithClusterAwareChecks(t *testing.T) {
...
+										Raw: []byte(`4`),
+									},
+								}).
+							Build()).
+					Build(),
+			},
+			oldClusterClass: clusterClassBuilder.
+				WithVariables(
+					clusterv1.ClusterClassVariable{
+						Name: "zone",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1362]:
```suggestion
						Name: "cpu",
```
same in l.1375

The test currently passes because there is no schema validation for the cpu variable.

**Review by sbueringer — COMMENTED** [review]
To summarize open points from my side:
* possibly we want to return early if the new ClusterClass schema is invalid to avoid flooding users with errors: https://github.com/kubernetes-sigs/cluster-api/pull/5752#discussion_r761845903
* we can potentially simplify schemasAreEqual: https://github.com/kubernetes-sigs/cluster-api/pull/5752#discussion_r761842371

+ two small new nits

@fabriziopandini If you have time, it would be great if you can take another look.

File: internal/webhooks/clusterclass.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: @fabriziopandini WDYT?

File: internal/webhooks/clusterclass.go (hunk 20 above)
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Elegant solution :). I somehow didn't think of that one :D

File: internal/webhooks/clusterclass.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go:371]: Debugging reflect.DeepEqual is very painful :/

File: internal/webhooks/clusterclass.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Can be resolved, thx!!

File: internal/webhooks/clusterclass.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: Honestly not exactly sure what I was doing before when I thought this wasn't working - but it's working now 😆 

I've updated the code to use the DeepEqual.

**Comment by sbueringer** [comment]
lgtm pending squash from my side

The only open conversation is:
* possibly we want to return early if the new ClusterClass schema is invalid to avoid flooding users with errors: https://github.com/kubernetes-sigs/cluster-api/pull/5752#discussion_r761845903

I think it would be nice to hear opinions on that, but either way it's not a blocker for me.

File: internal/webhooks/clusterclass.go (hunk 8 above)
Comment (vincepri) [reply internal/webhooks/clusterclass.go:252]: Something to think about is to not use specific fields to signal the action taken on it, for example today we have `altered`, `added`, `removed`, `required`. These could be kinds of errors which we might enumerate somewhere else and then have a single aggregator struct that when the error is added works on the enumeration rather than going list by list.

Does that make sense? Happy to expand further

File: internal/webhooks/clusterclass.go [hunk 22]
```diff
@@ -154,20 +163,210 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
...
+// - added which have been added to the ClusterClass on update.
+// - altered which have had their schemas changed on update.
+// - removed which have been removed from the ClusterClass on update.
+// - required (with 'required' : true) from the new ClusterClass.
+func getClusterClassVariablesForValidation(oldVars, newVars map[string]*clusterv1.ClusterClassVariable) clusterClassVariablesForValidation {
+	// Create a map each for new, altered,  removed and required Variables.
+	addedVariables := map[string]*clusterv1.ClusterClassVariable{}
+	alteredVariables := map[string]*clusterv1.ClusterClassVariable{}
+	removedVariables := map[string]*clusterv1.ClusterClassVariable{}
+	requiredVariables := map[string]*clusterv1.ClusterClassVariable{}
```
Comment (vincepri) [inline internal/webhooks/clusterclass.go:342]: Same as above, consider having the action type of the variable defined somewhere else rather than carrying around multiple maps.

We should try to avoid having "business logic"/semantic be part of the variable name, but rather have the semantic behavior as an input

File: cmd/clusterctl/cmd/describe_cluster.go [hunk 23]
```diff
@@ -114,7 +114,7 @@ func init() {
 		"Groups machines when ready condition has the same Status, Severity and Reason.")
 	describeClusterClusterCmd.Flags().BoolVar(&dc.disableGrouping, "disable-grouping", false,
 		"Disable grouping machines when ready condition has the same Status, Severity and Reason.")
-	describeClusterClusterCmd.Flags().MarkDeprecated("disable-grouping",
+	_ = describeClusterClusterCmd.Flags().MarkDeprecated("disable-grouping",
```
Comment (fabriziopandini) [inline cmd/clusterctl/cmd/describe_cluster.go:117]: This should now be fixed on main

File: internal/webhooks/clusterclass.go (hunk 22 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: I've added a variableValidationKey as the map key here so this is now a single map where the key has the action type + variable name. This simplifies the logic a bit and gets rid of keeping the logic in the variable name - let me know what you think.

File: internal/webhooks/clusterclass.go (hunk 8 above)
Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: I've updated this now, but in a way that doesn't really serve to make it more generic (just removes all the extra maps) I'd like to punt on this and tackle the more generic error handling in a future PR WDYT?

File: internal/webhooks/clusterclass.go (hunk 22 above)
Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: To me the change has the following downsides:
* `added`, `altered`, `removed` do not sound like `variableValidationAction`'s to me. `variableValidationAction` sounds to me like it's the type of the `validation`, which does only fit for `required`. Or maybe it is actually the cause why we want to validate something, but then `variableValidationAction` sound strange to me.
* Previously the errors where ordered, i.e. all removed, all altered, all added all required errors. I think now it's random and thus also non-deterministic.
* The new code is harder to read and understand.

But maybe that's just me. If you'll get consensus on the current implementation, I don't want to block.

Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: Agreed that there are definite downsides to this approach. 
- The naming / additional struct aren't ideal - but this is the 'type' of vallidation being done to my mind: We have four validation paths and each of these keys leads to some validation action being performed based on the 'type'. To me there's no way to really separate the cause/action categorically here, but that leads to some confusion - have you another idea on naming? 

- I don't think the ordering is important due to the way errors are being aggregated though - we're still reporting the index in the error message, but I don't think we should care about the order we're printing those messages.

- Agreed this is probably harder to read, but gets rid of the issue with using specific map names to define actions. I'm happy to go either direction on the structure here.

@vincepri WDYT?

Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: > have you another idea on naming?

Best idea I can come up with is something like `validationType`, but that's just potentially better because it's more generic and thus less expressive / potentially less wrong :)

> I don't think the ordering is important due to the way errors are being aggregated though - we're still reporting the index in the error message, but I don't think we should care about the order we're printing those messages.

I just thought as a user faced with a wall of errors it might drive me crazy to get the different error types in mixed order and in a different order each time I run against the webhook (aka when I'm working my way through fixing those errors). Imagine golangci-lint gives you a long list of errors, each time in a different order and the types are mixed (maybe that's the current behavior in fact :P).

Comment (killianmuldoon) [reply internal/webhooks/clusterclass.go]: I hadn't thought of it that way - I've made an update to make the error order defined (in terms of error type, variable impacted is still unordered inside that category). I've also update the name of the to variableValidationType to be more generic.

Comment (sbueringer) [reply internal/webhooks/clusterclass.go]: Looks good from my side.

**Comment by fabriziopandini** [comment]
Yay! This is the last PR of a really complex feature, thanks again for driving this effort! 
/lgtm
/approve

**Comment by sbueringer** [comment]
> Yay! This is the last PR of a really complex feature, thanks again for driving this effort! /lgtm /approve

Great work!
